### ISSUE TITLE
FirebaseIndexRecyclerAdapter got broken on 2.0.1 if not all data have permission

### ISSUE THREAD

**Issue by Shirane85** [firebase/FirebaseUI-Android#761]
* Android device: _____
  * Android OS version: _____
  * Google Play Services version: _____
  * Firebase/Play Services SDK version: 11.0.1
  * FirebaseUI version: 2.0.1

I have two tables: Users and Items.
User have a set of keys of the items he shared with. Therefore I used `FirebaseIndexRecyclerAdapter` to display them.
There is a situation when user is being un-shared to an item and then he doesn't have permission to GET it (using Firebase DB rules). 
On 1.2.0 the `FirebaseIndexRecyclerAdapter` just ignored those items and display the rest but now i get `onCancelled` called and nothing is displayed.

#### Steps to reproduce:

  1. Use `FirebaseIndexRecyclerAdapter`
  2. Make one of the items return "Permission denied" when trying to get it
  
#### Observed Results:

  `onCancelled` is called and nothing gets displayed
  
#### Expected Results:

`FirebaseIndexRecyclerAdapter` should ignore the non permission items and display the rest (as it was on 1.2.0)

**Comment by SUPERCILEX** [comment]
@Shirane85 Is it the keys that lack permission or the actual data? In neither case do we stop listening for data if one item fails, but the core Firebase SDK might have changed its behavior regarding populating the keys. If that's the case, there's not much FirebaseUI can do.

**Comment by puf** [comment]
@SUPERCILEX I don't think anything changed in the way the Firebase SDKs deal with listening for unauthorized data. Do you have a pointer for what makes you say that?

**Comment by SUPERCILEX** [comment]
@puf no, that was just the only thing I could think of because we don't do anything in our `onCancelled` methods but log stuff.

Here are the places where we handle `onCancelled`:
- https://github.com/firebase/FirebaseUI-Android/blob/master/database/src/main/java/com/firebase/ui/database/FirebaseArray.java#L144
- https://github.com/firebase/FirebaseUI-Android/blob/master/database/src/main/java/com/firebase/ui/database/FirebaseIndexArray.java#L121
- https://github.com/firebase/FirebaseUI-Android/blob/master/database/src/main/java/com/firebase/ui/database/FirebaseIndexArray.java#L266
- https://github.com/firebase/FirebaseUI-Android/blob/master/database/src/main/java/com/firebase/ui/database/FirebaseRecyclerAdapter.java#L122

So basically, I don't see how this issue is possible.

**Comment by puf** [comment]
Is it possible that those calls to `notifyListenersOnCancelled()` could end up not showing anywhere if the developer didn't attach a listener? If so, make sure the library also logs them (as a warning).

Btw: if I understand the issue correctly, it occurs when the actual data becomes inaccessible while the code has already read it before. That might help in being able to reproduce it.

@Shirane85 it would be really helpful if you could provide an [MCVE](http://stackoverflow.com/help/mcve).

**Comment by exim13** [comment]
I actually just ran into this issue today. I had to downgrade Firebase to 10.2.0 from 11.0.2 and Firebase UI to 1.2.0 from 2.0.1.

In my case, I have FirebaseIndexRecyclerAdapter in a fragment. Another fragment is used to remove an item. Upon popping the backstack, no items appears anymore. It occasionally displays a warning saying I don't have permission to view the deleted item, thus leaving the entire recyclerview blank.

Going back to 10.2.0 in Firebase and 1.2.0 in Firebase UI solved the problem. Once the backstack is popped, the items are correctly shown, and the item removed is gone.

**Comment by exim13** [comment]
So downgrading firebase was only a temporary fix. The problem started again without changing anything in the app or database rules, even with the downgrade.

So I guess my problem lies elsewhere. It's weird since only removing items while in another fragment seems to be the problem. I tested removing an item without the use of an additional fragment on top and the FirebaseIndexRecyclerAdapter handled it properly. There was an onCancelled error due to permissions, but the item was removed from the view.

With that additional fragment, I get an empty view and no onCancelled error of any kind. My onCreateView calls for a new FirebaseIndexRecyclerAdapter when the backstack is popped and has no problems with other fragments on top. My onDestroyView calls for cleanup() on the list adapter. I have tried turning setPersistenceEnabled(false) also, which didn't affect my results. I have been stumped haha

**Comment by SUPERCILEX** [comment]
@exim13 I'd be willing to try and help you figure it out if you can share your code.

**Comment by exim13** [comment]
Thanks @SUPERCILEX ! Wasn't sure how much code to include, but this is everything pertaining to the FirebaseIndexRecyclerAdapter. Let me know if you need anything else!

Firebase structure:
`/leagues/$league_id/ league info here`
`/users/$user_id/leagues/ list of user's leagues here as $league_id: true`

Fragment displaying leagues:
```java
public class Leagues extends Fragment {
    private RecyclerView recyclerView;
    private FirebaseIndexRecyclerAdapter listAdapter;
    private LinearLayoutManager linearLayoutManager;
    private TextView emptyText;

    private FirebaseDatabase firebaseDatabase;
    private DatabaseReference leagueDatabaseReferenece;
    private DatabaseReference userDatabaseReference;
    private FirebaseAuth firebaseAuth;

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        firebaseAuth = FirebaseAuth.getInstance();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.league_fragment, container, false);

        recyclerView = (RecyclerView) v.findViewById(R.id.league_list);
        emptyText = (TextView)v.findViewById(R.id.empty_league_list_text);
        linearLayoutManager = new LinearLayoutManager(getActivity());
        recyclerView.setLayoutManager(linearLayoutManager);

        FirebaseUser user = firebaseAuth.getCurrentUser();
        if(user != null) {
            final ProgressDialog dialog = new ProgressDialog(getActivity());
            dialog.setMessage("Loading Leagues");
            dialog.show();
            firebaseDatabase = FirebaseDatabaseUtil.getDatabase();
            userDatabaseReference = firebaseDatabase.getReference().child("users")
                    .child(user.getUid()).child("leagues");
            leagueDatabaseReferenece = firebaseDatabase.getReference().child("leagues");
            listAdapter = new FirebaseIndexRecyclerAdapter<League, LeagueHolder>(League.class,
                    R.layout.league_list_row, LeagueHolder.class,
                    userDatabaseReference, leagueDatabaseReferenece){
                @Override
                protected void populateViewHolder(LeagueHolder holder, League league, int position) {
                    holder.setLeagueTitle(league.getTitle());
                }

                @Override
                public LeagueHolder onCreateViewHolder(ViewGroup parent, int viewType) {
                    LeagueHolder viewHolder = super.onCreateViewHolder(parent, viewType);
                    viewHolder.setOnClickListener(new LeagueHolder.ClickListener() {
                        @Override
                        public void onItemClick(View view, int position) {
                            showScores(listAdapter.getRef(position).getKey(), view);
                        }
                    });
                    return viewHolder;
                }

                @Override
                public void onDataChanged() {
                    super.onDataChanged();
                    if (dialog != null && dialog.isShowing()) {
                        dialog.dismiss();
                    }
                    emptyText.setVisibility(getItemCount() == 0 ? View.VISIBLE : View.GONE);
                }

                @Override
                public void onChildChanged(EventType type, DataSnapshot snapshot, int index, int oldIndex) {
                    emptyText.setVisibility(getItemCount() == 0 ? View.VISIBLE : View.GONE);
                    super.onChildChanged(type, snapshot, index, oldIndex);
                }

                @Override
                public void onCancelled(DatabaseError error) {
                    super.onCancelled(error);
                }
            };
            DividerItemDecoration dividerItemDecoration = new DividerItemDecoration(recyclerView.getContext(),
                    linearLayoutManager.getOrientation());
            recyclerView.addItemDecoration(dividerItemDecoration);
            recyclerView.setAdapter(listAdapter);
        }
        getActivity().invalidateOptionsMenu();
        return v;
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        if(listAdapter !=null) listAdapter.cleanup();
    }

    public static class LeagueHolder extends RecyclerView.ViewHolder{
        private final TextView leagueTitleTextView;
        private final LinearLayout leagueLayout;

        public LeagueHolder(View itemView) {
            super(itemView);
            leagueTitleTextView = (TextView) itemView.findViewById(R.id.league_title);
            leagueLayout = (LinearLayout)itemView.findViewById(R.id.league_layout);
            leagueLayout.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    mClickListener.onItemClick(v, getAdapterPosition());
                }
            });
        }

        private LeagueHolder.ClickListener mClickListener;
        //Interface to send callbacks...
        public interface ClickListener{
            void onItemClick(View view, int position);
        }

        public void setOnClickListener(LeagueHolder.ClickListener clickListener){
            mClickListener = clickListener;
        }

        public void setLeagueTitle(String leagueTitle){
            leagueTitleTextView.setText(leagueTitle);
        }
    }

    private void showScores(String firebaseLeagueId, View v){
		// attach arguments to send to Scores fragment
		...
		
        getActivity().invalidateOptionsMenu();
        Fragment fragment = new Scores();
        FragmentManager fragmentManager = getFragmentManager();
        int fragmentPosition = R.id.left_fragment;
        fragmentManager.beginTransaction()
                .addToBackStack(null)
                .replace(fragmentPosition, fragment)
                .commit();
    }
}
```

Part of fragment that removes user from league:
```java
private void leaveLeague(){
        FirebaseUser user = firebaseAuth.getCurrentUser();
        if(user != null){
            // Update user count
            DatabaseReference leaguesDatabaseRef = firebaseDatabase.getReference().child("leagues").child(firebaseLeagueId).child("userCount");
            leaguesDatabaseRef.setValue(userCount - 1);

			// Remove league from user
            DatabaseReference databaseRef = firebaseDatabase.getReference().child("users")
                    .child(user.getUid()).child("leagues");
            databaseRef.child(firebaseLeagueId).removeValue();

            FragmentManager manager = getFragmentManager();
            if (manager.getBackStackEntryCount() > 0) {
                FragmentManager.BackStackEntry first = manager.getBackStackEntryAt(0);
                manager.popBackStack(first.getId(), FragmentManager.POP_BACK_STACK_INCLUSIVE);
            }
            getActivity().invalidateOptionsMenu();

            Fragment fragment = new Leagues();
            FragmentManager fragmentManager = getFragmentManager();
            fragmentManager.beginTransaction()
                    .addToBackStack(null)
                    .replace(R.id.left_fragment, fragment)
                    .commit();
        }
    }
```

**Comment by SUPERCILEX** [comment]
@exim13 Hmmm, yeah I don't see anything wrong with your adapter code which makes me think the only thing going wrong would be your fragment transactions. Since I'm not familiar with the flow of your app I can't really help you much, but I would recommend watching [this I/O talk](https://youtu.be/eUG3VWnXFtg?t=2m52s) about fragments and the back stack since what you're doing looks a little weird to me. Sorry I couldn't offer more help! 😊

**Comment by exim13** [comment]
Thank you @SUPERCILEX for taking a look at it. I was never really confident with fragment transactions. and will definitely check out the video. Hopefully that is the problem! Thanks again for the help and sorry to take over this issue with a totally separate problem.

**Comment by SUPERCILEX** [comment]
@exim13 No problem! 😀

**Comment by samtstern** [comment]
@SUPERCILEX is there still an issue here?

**Comment by SUPERCILEX** [comment]
@samtstern I don't believe so! 😀
